File: apps/ios/Packages/StoryArcKit/Sources/Playback/Audiobook.swift
public import Foundation

/// One part of an audiobook: a chapter marker inside a file, or a whole file of a folder.
///
/// `design.md`'s table calls both a *part*, and `publication-formats` requires them to
/// behave the same way — "its parts — the files, or the whole of a single file — stand in
/// for chapters". So one type, carrying where the audio is and where inside it the part
/// begins.
///
/// **Here rather than in `Formats`, and the precedent is `PublicationFormat`'s.** That enum
/// moved to `StoryArcCore` because the library sorts, filters and explains by format and
/// none of that should require the parser. The same holds one layer along: the player draws
/// parts, seeks between them and names them, and none of that should require a ZIP reader.
/// `Formats` produces these; it does not own them.
public struct AudiobookPart: Sendable, Equatable {
    /// The file this part is in. The same URL for every chapter of an M4B; a different one
    /// for every part of a folder.
    public let url: URL

    /// What the container calls it, or `nil` when it names nothing.
    ///
    /// **Never the file name.** `design.md` records that as a product decision:
    /// `01 - track.mp3` is not what a listener is in the middle of. A part with no name is
    /// numbered by ``PlayerCentre``, which is the one place that decision lives.
    public let title: String?

    /// Where this part starts inside ``url``. Zero for a whole file.
    public let start: TimeInterval

    /// How long it runs, when the container says.
    public let duration: TimeInterval?

    public init(url: URL, title: String?, start: TimeInterval, duration: TimeInterval?) {
        self.url = url
        self.title = title
        self.start = start
        self.duration = duration
    }

    /// Where this part ends inside its file, when its length is known.
    public var end: TimeInterval? { duration.map { start + $0 } }
}

/// An audiobook as the format layer sees it: parts in playing order, and what was lost.
public struct Audiobook: Sendable, Equatable {
    public let parts: [AudiobookPart]

    /// How many entries looked like parts and could not be read.
    ///
    /// `publication-formats`: a damaged audiobook "plays what it can and states how much it
    /// could not … the count is stated in the player's own controls rather than
    /// interrupting playback" — the same rule that opens a comic missing pages.
    public let unreadablePartCount: Int

    public init(parts: [AudiobookPart], unreadablePartCount: Int) {
        self.parts = parts
        self.unreadablePartCount = unreadablePartCount
    }

    /// The whole book's length, or `nil` when any part's is unknown.
    public var duration: TimeInterval? {
        let durations = parts.compactMap(\.duration)
        guard durations.count == parts.count, !parts.isEmpty else { return nil }
        return durations.reduce(0, +)
    }
}

File: apps/android/core/playback/src/main/kotlin/app/storyarc/core/playback/Audiobook.kt
package app.storyarc.core.playback

/**
 * A publication to play, and where its audio is.
 *
 * The whole of what the format layer has to hand the player. Deliberately not a
 * `Publication`: `:core:playback` decodes audio and has no business knowing that a library
 * sorts by series, and a service rebuilt after process death has a URI and a title and
 * nothing else. The app assembles one of these from a `Publication` when it starts a book.
 */
data class Audiobook(
    /** The publication's stable id, so a position resolves through content identity. */
    val id: String,
    val title: String,
    val author: String? = null,
    /**
     * The audio, in playing order.
     *
     * One entry for a single file — chaptered or not — and one per file for a folder.
     * `publication-formats` puts a folder's parts in the same natural order that makes a
     * folder of images one comic; the ordering happened in the format layer and this is
     * the result of it, not a second chance to reorder.
     */
    val sources: List<AudioPart>,
    /** Cover art to put on the lock screen, as a URI the platform can fetch. */
    val artworkUri: String? = null,
    /**
     * Parts that were found and cannot be played.
     *
     * `publication-formats`: a damaged audiobook plays "what it can and states how much it
     * could not … in the player's own controls rather than interrupting playback".
     */
    val skippedPartCount: Int = 0,
) {

    /** Whether the parts are separate files or chapter marks inside one. */
    val layout: PartLayout
        get() = if (sources.size > 1) PartLayout.FILES else PartLayout.MARKS

    /** One playable file. */
    data class AudioPart(
        /** Where the bytes are, as a URI string the platform's data source understands. */
        val uri: String,
        /** What a listener is told they are in the middle of, for a folder's parts. */
        val title: String,
    )
}

/**
 * How a publication's parts map onto what the decoder plays.
 *
 * The one thing the player has to branch on, and it is worth a type rather than a boolean
 * because the two cases move the audio in genuinely different ways: choosing a part is a
 * change of item in one and a seek in the other, and a wrong branch is a chapter list
 * whose rows go to the wrong place.
 */
enum class PartLayout {
    /** A folder. Each part is its own item, and the decoder moves between them. */
    FILES,

    /**
     * A single file. Every part is a mark inside one item, so choosing one is a seek and
     * the current part has to be worked out from the position.
     */
    MARKS,
}
